File: sentimental-credit/credit.py
from cs50 import get_int

# Luhn's algorithm


def checkvalid(cardnum, len):
    sum_even = 0
    sum_odd = 0
    for i in range(1, len+1):
        # extract digits
        current_digit = int((cardnum / pow(10, i-1)) % 10)
        # for even numbers
        if i % 2 == 0:
            product_even = current_digit * 2
            if (product_even < 10):
                # add digit
                sum_even += product_even
            else:
                # if digit * 2 > 10, add individual digits
                sum_even += (product_even - 9)
        # for odd numbers
        else:
            sum_odd += current_digit
        sum = sum_even + sum_odd
    # if sum ends with 0, valid
    if sum % 10 == 0:
        return True
    # if sum doesn't end with 0, invalid
    else:
        return False


def checkdigit(cardnum, digits):
    checkdigits = cardnum // pow(10, len(str(card_number))-digits)
    return checkdigits


# get card number
card_number = get_int("Number: ")
# check if length is valid: 16, 15 or 13 digits only
card_len = len(str(card_number))

if card_len != 16 and card_len != 15 and card_len != 13:
    print("INVALID\n")

# if length = 15, check first two digits for AMEX (34 or 37)
if card_len == 15:
    first_digits = checkdigit(card_number, 2)
    if first_digits != 34 and first_digits != 37:
        print("INVALID\n")
    else:
        valid = checkvalid(card_number, 15)
        if valid:
            print("AMEX\n")
        else:
            print("INVALID\n")


# if 13, check first digit for visa
    # then check valid
if card_len == 13:
    first_digits = checkdigit(card_number, 1)
    print(f"first digit: {first_digits}")
    if first_digits != 4:
        print("INVALID\n")
    else:
        valid = checkvalid(card_number, 13)
        if valid:
            print("VISA\n")
        else:
            print("INVALID\n")

if card_len == 16:
    # check digit for visa
    first_digits = checkdigit(card_number, 1)
    if checkdigit(card_number, 1) == 4:
        valid = checkvalid(card_number, 16)
        if valid:
            print("VISA\n")
        else:
            print("INVALID\n")
    elif checkdigit(card_number, 2) > 50 and checkdigit(card_number, 2) < 56:
        valid = checkvalid(card_number, 16)
        if valid:
            print("MASTERCARD\n")
        else:
            print("INVALID\n")
    else:
        print("INVALID\n")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <math.h>

int numdgt;
int getdigit(long crdnum, int numbit);
int crdtype;
int checkvalid(long crdnum);
int currentdigit;
long power(int exponent);
int prdeven;
int sumeven;
int sumodd;
int sum;
int result;

int main(void)
{
// get card number
    long crdnum = get_long("Number: ");
// check length of card number
    numdgt = floor(log10(crdnum))+1;
// check if length is valid - 16, 15 or 13
    if (numdgt != 16 & numdgt != 15 & numdgt != 13)
    {
        printf("INVALID\n");
    }
//first two digits for AMEX
    if (numdgt == 15)
    {
        crdtype = crdnum / pow(10,(numdgt-2));
        if (crdtype !=34 & crdtype!=37)
        {
            printf("INVALID\n");
        }
        else
        {
            result = checkvalid(crdnum);
            if (result == 1)
            {
                printf("AMEX\n");
            }
            else printf("INVALID\n");
        }
    }

//first digit for VISA 13 digit card
    if (numdgt == 13)
    {
        crdtype = crdnum / pow(10,(numdgt-1));
        if (crdtype !=4 & crdtype!=4)
        {
            printf("INVALID\n");
        }
        else
        {
            result = checkvalid(crdnum);
            if (result == 1)
            {
                printf("VISA\n");
            }
            else printf("INVALID\n");
        }

    }
//first digit for VISA 16 digit card
    if (numdgt == 16)
    {
        crdtype = crdnum / pow(10,(numdgt-1));
        if (crdtype == 4)
        {
            result = checkvalid(crdnum);
            if (result == 1)
            {
                printf("VISA\n");
            }
            else printf("INVALID\n");
        }
        else
        {
        crdtype = crdnum / pow(10,(numdgt-2));
        if (crdtype > 50 & crdtype < 56)
        {
            result = checkvalid(crdnum);
            if (result == 1)
            {
                printf("MASTERCARD\n");
            }
            else printf("INVALID\n");
        }
        else printf("INVALID\n");
        }
    }
}


//check Luhn's Algorithm, defining checkvalid function
int checkvalid(long crdnum)
{
 for(int i = 1; i <= 16; i++)
    {
        // extract digits
    currentdigit = ((crdnum/power(i-1))%10);
    //for even numbers
    if (i % 2 == 0)
    {
        // multiply digit by 2
        prdeven = currentdigit * 2;
        if (prdeven < 10)
        {
            sumeven += prdeven;
        }
        else
        {
            sumeven += (prdeven-9);
        }
    }
    // for odd numbers
    else
    {
        sumodd += currentdigit;
    }
    // add all digits
    sum = sumeven + sumodd;
    }
    if (sum%10==0)
        return 1;
    else return 0;
}

long power(int exponent)
{
    return pow(10, exponent);
}